enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,435 @@
+
+### API Extensions
+
+- Adds a top-level PlatformOperator cluster-scoped CRD.
+- Adds the Bundle and BundleDeployment OLM 1.x API(s).
+- Adds a validating webhook for the Bundle and BundleDeployment APIs.
+- Adds a CRD validation webhook.
+
+### Risks and Mitigations
+
+- Early phases will introduce multiple v1alpha1 CRDs downstream.
+- Early phases will introduce OLM v1 components that are in alpha stages.
+- No concrete migration plan from legacy to OLM v1.
+- Coordination around external team dependencies.
+- The BundleDeployment API has a large net of permissions that may require upfront RBAC to limit this API to admins.
+
+### Drawbacks
+
+- Implementing this PO mechanism using legacy OLM would provide more immediate business value. That implementation is outlined below in the alternatives section.
+- Early phases will deliver through tech preview guidelines. This leads to little customer feedback, given it precludes production clusters.
+- Early phases may introduce a new OLM 1.x control plane that has little-to-no interaction with the existing, legacy OLM 0.x control plane.
+- Early phases may introduce validating webhooks. These webhooks can impact the cluster's overall health when they're not available.
+- Early phases may introduce validating webhooks. By default, these webhooks will be deployed using an HA configuration, but may not have logic to accommodate SNO cluster environments.
+- Early phases may introduce a dynamic watching implementation for managing the underlying OLM operator resources. This has potential performance implications that need to be measured through benchmarking and/or addressed before any GA phase.
+- Future phase designs may require a large scale rewrite of the OLM component.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Does the console need to be able to provide a UI experience for POs?
+- Does having separate APIs for ClusterOperators and PlatformOperators provide a degraded admin UX?
+- How to handle the maxOpenShiftVersion constraint given it's defined in the CSV resource?
+- Dedicated CatalogSource? Dedicated channel? The top-level PO API contains a catalog configuration?
+- Is it possible to have a dedicated PO index image?
+- Should the existing minKubeVersion and maxOpenShiftVersion fields be unified?
+- How are phase 0 requirements enforced?
+- Does this phase need to enforce cluster invariants?
+- What legacy OLM runtime changes are needed?
+- How to prevent an admin installing an operator that was installed through the PO mechanism?
+- How do admins determine the PO package name ahead of time?
+- What's the desired upgrade workflow and admin UX?
+- Are cluster admins allowed to pin package versions?
```

Comment:
yes

Reply:
Does that imply that we'd need a version/versionRange field for any top-level API in subsequent phases? How would an admin determine this version information upfront before cluster creation, or do we imagine that the use case is editing any top-level API after cluster creation, and at runtime?

Reply:
> How would an admin determine this version information upfront before cluster creation, or do we imagine that the use case is editing any top-level API after cluster creation, and at runtime?

i expect it's more often going to be a post-install thing where they need to pin to a version to get a particular fix/avoid a regression.

but at install time they may know from existing clusters what version they want to install, and they may want to codify that into a gitops flow that ensures every cluster that's stamped out gets the exact same versions of the POs.

or they get the version of the operator product's release notes?

Reply:
> Is there a reason not to give admins the same level of control of versions of platform operators that they have for the underlying platform? 

we need to give them slightly more (ability to upgrade/change the version of a PO independent of the cluster) to address the fact that the POs may version/release on their own cadences so an admin may want to pick up a new version even when not upgrading their whole cluster, or may need to revert to an older version.

Whether we need to support automated updates outside of a cluster upgrade, perhaps not.  But supporting them is largely trivial, once we have to support manual updates and "with the cluster" updates anyway.  But automated updates outside of a cluster upgrade can definitely be a phase 3 thing.

Reply:
I'm not sure we're communicating 100%.

I thought one aspect of POs was that they upgraded when OCP did. That is the only case where they should change version without an explicit request from the user tied to the PO.

Reply:
I'm going to consolidate this thread with the [original cluster lifecycle thread](https://github.com/openshift/enhancements/pull/1170#discussion_r915034676) to hopefully reduce the amount of UI strain of opening this PR.